Angela Lin - Feed API Challenge

tigclaw opened the pull request.

danillahtin commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -19,6 +19,14 @@ public final class RemoteFeedLoader: FeedLoader {
+		client.get(from: url) { [weak self] result in
+			guard self != nil else { return }
+			switch result {
+			case .success((let data, let response)):
```
The `let` can be defined outside the tuple so the expression can be shorter:
```suggestion
			case let .success((data, response)):
```

tigclaw replied:
Cool, I'll have to remember this in the future.

danillahtin commented on FeedAPIChallenge/Feed API/FeedImageMapper.swift:
```diff
@@ -0,0 +1,47 @@
+
+import Foundation
+
+internal final class FeedImageMapper {
```
`internal` is the default access level, so you don't need to write it explicitly.

It was written explicitly in the first lectures of the program for teaching purposes, but you don't need to do it in your code.

Please, remove all explicit `internal` keywords.
```suggestion
final class FeedImageMapper {
```
Suggestion:
`FeedImageMapper` is defined as a class, so we can instantiate it as `FeedImageMapper()`. However, all its methods are static - it has no instance methods or properties.

It may be confusing for other developers to be able to instantiate a class but then not be able to do anything with it.

As you'll see in the following lectures, you can make the initializer private in such cases since the `FeedImageMapper` is just a namespace for static methods, with no instance-specific behavior.

Or you can make it an enum with no cases, which can never be instantiated!

Both solutions will disallow instantiation at compile time so that `FeedImageMapper()` won't compile ✅
```suggestion
enum FeedImageMapper {
```

danillahtin replied:
There's no standard way here. Using an enum is my own preference as it is very short and doesn't require explicitly making the initializer private. If you feel like using structs or classes with private initializers it is absolutely ok.

danillahtin commented on FeedAPIChallenge/Feed API/FeedImageMapper.swift:
```diff
@@ -0,0 +1,47 @@
+	private static var OK_200: Int { 200 }
+
+	internal static func map(_ data: Data, _ response: HTTPURLResponse) -> FeedLoader.Result {
+		guard response.statusCode == OK_200 else {
+			return .failure(RemoteFeedLoader.Error.invalidData)
+		}
+		do {
+			let root = try JSONDecoder().decode(NetworkFeedImageContainer.self, from: data)
+			return .success(root.feedImages)
+		} catch {
+			return .failure(RemoteFeedLoader.Error.invalidData)
+		}
```
Since the error inside `catch` block is ignored optional `try?` can be used.
```swift
guard response.statusCode == OK_200 else {
	return .failure(RemoteFeedLoader.Error.invalidData)
}

guard let root = try? JSONDecoder().decode(NetworkFeedImageContainer.self, from: data) else {
	return .failure(RemoteFeedLoader.Error.invalidData)
}

return .success(root.feedImages)
```

And now it becomes clear that two guard blocks can be combined into one to remove duplication of `return .failure(RemoteFeedLoader.Error.invalidData)`:
```suggestion
		guard response.statusCode == OK_200, let root = try? JSONDecoder().decode(NetworkFeedImageContainer.self, from: data) else {
			return .failure(RemoteFeedLoader.Error.invalidData)
		}

		return .success(root.feedImages)
```

danillahtin commented on FeedAPIChallenge/Feed API/FeedImageMapper.swift:
```diff
@@ -0,0 +1,47 @@
+		}
+	}
+
+	private struct NetworkFeedImageContainer: Decodable {
```
Suggestion:
Since this struct is private, there's no need to give extra context in the name of the type. It can be simplified to just `Container`
```suggestion
	private struct Container: Decodable {
```

danillahtin reviewed (commented):
Hi @tigclaw, you almost passed the challenge!

This is Danil, I am an instructor at Essential Developer.

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

tigclaw commented:
> Please, check my review comments and let me know once you're done so I can review it again.

Hi Danil @danillahtin  - thanks for reviewing! I've addressed your comments in latest commit.

danillahtin reviewed (commented):
Hi @tigclaw,

Congratulations, you've passed the first challenge and got your first belt stripe ⚪️①!

✅ All tests are passing, so your implementation complies with the specs.

**[Update your status](https://iosleadessentials.slack.com/archives/CTAR0LVMM/p1629389971188200) in the community now with the following emoji to celebrate your achievement: `:white-belt-1:`**

Great job, and keep it up! Remember, we're here to help you progress and get the most out of this program.

See you in the community and in the following challenge 🙂
